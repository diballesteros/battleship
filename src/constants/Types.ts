export type GameState = {
  playerShips: Ship[];
  computerShips: Ship[];
  computerMoves: number[];
  lastSuccessfulMoves: number[];
  successfulComputerHit: boolean;
  playerMoves: number[];
  completedComputerMoves: number[];
  gameResolved: boolean;
};

export type GameActions =
  | {
      type: 'SET_BOARD';
      payload: {
        playerShips: Array<Ship>;
        computerShips: Array<Ship>;
      };
    }
  | {
      type: 'RESET_BOARD';
    }
  | {
      type: 'PLAYER_MOVE';
      payload: {
        computerShips: Array<Ship>;
        playerMoves: Array<number>;
        completedComputerMoves: Array<number>;
        playerShips: Array<Ship>;
        gameResolved: boolean;
      };
    }
  | {
      type: 'COMPUTER_MOVE';
      payload: {
        computerMoves: Array<number>;
      };
    }
  | {
      type: 'UPDATE_HIT';
      payload: {
        lastSuccesfulMoves: Array<number>;
        successfulComputerHit: boolean;
      };
    };

export type Ship = {
  hits: boolean[];
  id: number;
  positions: number[];
};
