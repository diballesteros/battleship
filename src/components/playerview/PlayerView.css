.player_console {
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-around;
    align-items: center;
    height: 100%;
}

.player_board {
    display: flex;
    flex-flow: column nowrap;
    align-items: center;
}

.player_console label {
    color: #FFFFFF;
    font-size: 24px;
    margin: 10px;
}