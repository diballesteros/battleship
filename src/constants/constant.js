//Game
export const FIRSTCOLUMNSQUARES = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90];
export const LASTCOLUMNSQUARES = [9, 19, 29, 39, 49, 59, 69, 79, 89, 99];

//Gameboard.js
export const LETTERROW = [' ', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'];
export const NUMBERCOLUMN = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];

//ShipFactory.js
export const SHIPSTORE = [
    {
        model: 'Carrier',
        size: 5
    },
    {
        model: 'Battleship',
        size: 4
    },
    {
        model: 'Cruiser',
        size: 3
    },
    {
        model: 'Submarine',
        size: 3
    },
    {
        model: 'Destroyer',
        size: 2
    },
    {
        model: '',
        size: 0
    }
];

//Testing
export const TESTSHIPS = [
    {
        id: 1,
        positions: [1, 2, 3],
        hits: [false, false, false]
    },
    {
        id: 2,
        positions: [9, 19, 29],
        hits: [false, false, false]
    },
    {
        id: 3,
        positions: [42, 43, 44, 45, 46],
        hits: [false, false, false, false, false]
    },
    {
        id: 4,
        positions: [98, 99],
        hits: [false, false]
    },
    {
        id: 5,
        positions: [62, 72, 82, 92],
        hits: [false, false, false, false]
    }
];