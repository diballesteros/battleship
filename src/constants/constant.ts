import { DisplayState, GameState, TFactoryShip } from './Types';

// Game
export const FIRSTCOLUMNSQUARES = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90];
export const LASTCOLUMNSQUARES = [9, 19, 29, 39, 49, 59, 69, 79, 89, 99];

// Gameboard.js
export const LETTERROW = [' ', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'];
export const NUMBERCOLUMN = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];

// ShipFactory.js
export const SHIPSTORE: TFactoryShip[] = [
  {
    model: 'CARRIER',
    size: 5,
  },
  {
    model: 'BATTLESHIP',
    size: 4,
  },
  {
    model: 'CRUISER',
    size: 3,
  },
  {
    model: 'SUBMARINE',
    size: 3,
  },
  {
    model: 'DESTROYER',
    size: 2,
  },
];

// Testing
export const TESTSHIPS = [
  {
    id: 1,
    positions: [1, 2, 3],
    hits: [false, false, false],
  },
  {
    id: 2,
    positions: [9, 19, 29],
    hits: [false, false, false],
  },
  {
    id: 3,
    positions: [42, 43, 44, 45, 46],
    hits: [false, false, false, false, false],
  },
  {
    id: 4,
    positions: [98, 99],
    hits: [false, false],
  },
  {
    id: 5,
    positions: [62, 72, 82, 92],
    hits: [false, false, false, false],
  },
];

export const INITIAL_STATE: GameState = {
  playerShips: [],
  computerShips: [],
  computerMoves: [...Array(100).keys()],
  lastSuccessfulMoves: [],
  successfulComputerHit: false,
  playerMoves: [],
  completedComputerMoves: [],
  gameResolved: false,
};

export const INITIAL_DISPLAY_STATE: DisplayState = {
  currentTab: 0,
  isVertical: true,
  previousTab: 0,
  squareHeight: 30,
  squareWidth: 30,
};
