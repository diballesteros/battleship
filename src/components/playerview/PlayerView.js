import React, { Component } from 'react';
import Gameboard from '../gameboard/Gameboard';
import './PlayerView.css';

class PlayerView extends Component {
    constructor(props) {
        super(props);
        this.state = {
            playerShips: [
                {
                    id: 1,
                    positions: [1, 2, 3],
                    hits: [false, false, false]
                },
                {
                    id: 2,
                    positions: [9, 19, 29],
                    hits: [false, false, false]
                },
                {
                    id: 3,
                    positions: [42, 43, 44, 45, 46],
                    hits: [false, false, false, false, false]
                },
                {
                    id: 4,
                    positions: [98, 99],
                    hits: [false, false]
                },
                {
                    id: 5,
                    positions: [62, 72, 82, 92],
                    hits: [false, false, false, false]
                }
            ],
            computerShips: [
                {
                    id: 1,
                    positions: [1, 2, 3],
                    hits: [false, false, false]
                },
                {
                    id: 2,
                    positions: [9, 19, 29],
                    hits: [false, false, false]
                },
                {
                    id: 3,
                    positions: [42, 43, 44, 45, 46],
                    hits: [false, false, false, false, false]
                },
                {
                    id: 4,
                    positions: [98, 99],
                    hits: [false, false]
                },
                {
                    id: 5,
                    positions: [62, 72, 82, 92],
                    hits: [false, false, false, false]
                }
            ],
            computerMove: null,
            computerTurn: false
        }
    }

    isSunk(positions) {
        const sunkStatus = positions.every(position => position === true);
        sunkStatus && console.log('You sunk my battleship!');
        if (sunkStatus && this.state.computerTurn === true) {
            this.props.resetSuccessfulHit();
        }
        return sunkStatus;
    }

    resolveBoardState(move, shipId, ships) {
        const shipIndex = ships.findIndex(ship => ship.id === shipId);
        const hitIndex = ships[shipIndex].positions.indexOf(move);
        ships[shipIndex].hits[hitIndex] = true;
        this.isSunk(ships[shipIndex].hits);

        return ships
    };

    receivePlayerAttack(shipId, playerMove) {
        const modifiedComputerShips = this.resolveBoardState(playerMove, shipId, this.state.computerShips.slice());
        this.setState({
            computerShips: modifiedComputerShips
        });
    };

    computerCallback() {
        const computerMove = this.props.getComputerMove();
        this.setState({
            computerTurn: true,
            computerMove: computerMove
        });
    };

    resolveComputerTurn(shipId) {
        let modifiedPlayerShips = this.state.playerShips.slice();
        if (shipId) {
            this.props.setSuccessfulHit(this.state.computerMove);
            modifiedPlayerShips = this.resolveBoardState(this.state.computerMove, shipId, modifiedPlayerShips);
        }
        this.setState({
            computerTurn: false,
            playerShips: modifiedPlayerShips
        });
    }

    render() {
        return (
            <div className='player_console'>
                <div className='player_board'>
                    <label>My Board</label>
                    <Gameboard
                        ships={this.state.playerShips}
                        myBoard={true}
                        computerTurn={this.state.computerTurn}
                        computerMove={this.state.computerMove}
                        resolveComputerTurn={(shipId) => this.resolveComputerTurn(shipId)} />
                </div>
                <div data-testid='1' className='player_board'>
                    <label>Opponent's Board</label>
                    <Gameboard
                        ships={this.state.computerShips}
                        myBoard={false}
                        receivePlayerAttack={(shipId, i) => this.receivePlayerAttack(shipId, i)}
                        computerCallback={() => this.computerCallback()} />
                </div>
            </div>
        );
    }
}

export default PlayerView;
