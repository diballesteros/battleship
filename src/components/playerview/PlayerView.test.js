import React from 'react';
import { render } from '@testing-library/react';
import PlayerView from './PlayerView';
import {TESTSHIPS} from '../../constants/constant';

describe('PlayerView Component', () => {
    test('Initialized Playerview', () => {
        const { asFragment } = render( <PlayerView receivePlayerAttack={() => mockFn()} 
        playerShips={TESTSHIPS} 
        playerMoves={[]}
        computerShips={TESTSHIPS}
        completedComputerMoves={[]} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
