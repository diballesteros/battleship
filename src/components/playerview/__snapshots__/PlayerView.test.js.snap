// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerView Component Initialized Playerview 1`] = `
<DocumentFragment>
  <div
    class="player_console"
  >
    <div
      class="player_board"
    >
      <label>
        My Board
      </label>
      <div>
        <div
          class="letter_row"
        >
          <div
            class="square grid-square"
          >
             
          </div>
          <div
            class="square grid-square"
          >
            A
          </div>
          <div
            class="square grid-square"
          >
            B
          </div>
          <div
            class="square grid-square"
          >
            C
          </div>
          <div
            class="square grid-square"
          >
            D
          </div>
          <div
            class="square grid-square"
          >
            E
          </div>
          <div
            class="square grid-square"
          >
            F
          </div>
          <div
            class="square grid-square"
          >
            G
          </div>
          <div
            class="square grid-square"
          >
            H
          </div>
          <div
            class="square grid-square"
          >
            I
          </div>
          <div
            class="square grid-square"
          >
            J
          </div>
        </div>
        <div
          class="lower_section"
        >
          <div
            class="number_column"
          >
            <div
              class="square grid-square"
            >
              0
            </div>
            <div
              class="square grid-square"
            >
              1
            </div>
            <div
              class="square grid-square"
            >
              2
            </div>
            <div
              class="square grid-square"
            >
              3
            </div>
            <div
              class="square grid-square"
            >
              4
            </div>
            <div
              class="square grid-square"
            >
              5
            </div>
            <div
              class="square grid-square"
            >
              6
            </div>
            <div
              class="square grid-square"
            >
              7
            </div>
            <div
              class="square grid-square"
            >
              8
            </div>
            <div
              class="square grid-square"
            >
              9
            </div>
          </div>
          <div
            class="board"
          >
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ocean-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
            <div
              class="square ship-square my_square"
            />
          </div>
        </div>
      </div>
      <label>
        Sunk Ships: 0 / 5
      </label>
    </div>
    <div
      class="player_board"
      data-testid="1"
    >
      <label>
        Opponent's Board
      </label>
      <div>
        <div
          class="letter_row"
        >
          <div
            class="square grid-square"
          >
             
          </div>
          <div
            class="square grid-square"
          >
            A
          </div>
          <div
            class="square grid-square"
          >
            B
          </div>
          <div
            class="square grid-square"
          >
            C
          </div>
          <div
            class="square grid-square"
          >
            D
          </div>
          <div
            class="square grid-square"
          >
            E
          </div>
          <div
            class="square grid-square"
          >
            F
          </div>
          <div
            class="square grid-square"
          >
            G
          </div>
          <div
            class="square grid-square"
          >
            H
          </div>
          <div
            class="square grid-square"
          >
            I
          </div>
          <div
            class="square grid-square"
          >
            J
          </div>
        </div>
        <div
          class="lower_section"
        >
          <div
            class="number_column"
          >
            <div
              class="square grid-square"
            >
              0
            </div>
            <div
              class="square grid-square"
            >
              1
            </div>
            <div
              class="square grid-square"
            >
              2
            </div>
            <div
              class="square grid-square"
            >
              3
            </div>
            <div
              class="square grid-square"
            >
              4
            </div>
            <div
              class="square grid-square"
            >
              5
            </div>
            <div
              class="square grid-square"
            >
              6
            </div>
            <div
              class="square grid-square"
            >
              7
            </div>
            <div
              class="square grid-square"
            >
              8
            </div>
            <div
              class="square grid-square"
            >
              9
            </div>
          </div>
          <div
            class="board"
          >
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
            <div
              class="square ocean-square "
            />
          </div>
        </div>
      </div>
      <label>
        Sunk ships: 0 / 5
      </label>
    </div>
  </div>
</DocumentFragment>
`;
